### PR TITLE
feat(kinesis): `grantRead` now allows the `ListShards` action and `grant` is now public

### LINKED ISSUE #3357: Kinesis Stream permissions missing ListShards
* **I'm submitting a ...**
  - [x] :books: construct library gap

* **What is the current behavior?**
`Stream.grantRead` doesn't grant the ListShards permission, which is useful for Kinesis Client Library v2 among other things.

Stream also doesn't offer a custom `.grant(Grantable, List<String>)` to allow an easy workaround.

### PR DESCRIPTION
**Pull request by reillykw**
### Commit Message

`grant()` method on the class is private, preventing custom actions on attached policy. This makes that public.

`grantRead()` API now has `ListShards`, `DescribeStreamSummary`, or `SubscribeToShard` permissions

`grantWrite()` API no longer has `DescribeStream` permissions as it has been replaced by `ListShards` for shard discovery.
Rationale: Align with the [Kinesis Producer Library](https://aws.amazon.com/about-aws/whats-new/2019/11/amazon-kinesis-producer-library-now-supports-listshards-api-for-efficient-scaling-of-producer-applications/)

Closes #3357

BREAKING CHANGE: 
`grantWrite()` API no longer has `DescribeStream` permissions as it has been replaced by `ListShards` for shard discovery

### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by reillykw** [comment]
@rix0rrr @iliapolo Would it be possible to get a review on this in the near term?

**Comment by iliapolo** [comment]
@reillykw Thanks for this PR!

Yes I will get to it soon promptly 👍

**Comment by iliapolo** [comment]
@reillykw You mentioned in the PR description that this adds `ListShards` to `grantWrite` as well - where exactly is this reflected in the change? 

(Also notice the changes I made to the PR description and title)

**Comment by iliapolo** [comment]
@reillykw Can you please add a small section in the package README about the grant methods? I just noticed we don't actually have it and it would be nice to list out the various actions we are allowing with those methods.

**Comment by reillykw** [comment]
I reverted that change locally as write doesn't need listShards for write capabilities. I'll add a section to the README soon, thanks.

**Comment by reillykw** [comment]
@iliapolo Added documentation to the README, let me know if that is sufficient.  Thanks!

File: packages/@aws-cdk/aws-kinesis/README.md [hunk 1]
```diff
@@ -48,3 +48,24 @@ const stream = new Stream(this, 'MyEncryptedStream', {
 
 assert(stream.encryptionKey === myKmsKey);
 ```
+
+### Granting IAM Permissions on Stream to Grantable
+
+Read and Write iam actions can be added to a iam.IGrantable. If the stream has an encryption key attributed, access 
+to encrypt and/or decrypt with the encryption key will also be added to the grantable. There are three instance methods 
+available on a Stream object:  
+*.grantWrite(grantee: iam.IGrantable)
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-kinesis/README.md]: Notice the formatting here is not correct, you need a space after the `*`. Have a look at the [rendered](https://github.com/reillykw/aws-cdk/tree/kinesis-stream-permissions-list-shards/packages/%40aws-cdk/aws-kinesis#granting-iam-permissions-on-stream-to-grantable) version, you'll see where the formatting fails.

Also, thats not exactly the content I had in mind. Here, you dont need to list the methods that a `Stream` exposes, we have API docs for that.

File: packages/@aws-cdk/aws-kinesis/README.md [hunk 2]
```diff
@@ -48,3 +48,24 @@ const stream = new Stream(this, 'MyEncryptedStream', {
...
+const grantable = new iam.Role(this, 'Role', {
+  assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
+  description: 'Example role...',
+}
+
+const stream = new Stream(this, 'MyEncryptedStream', {
+    encryption: StreamEncryption.Kms
+});
+
+stream.grantRead(grantable);
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-kinesis/README.md]: This is good, but i also wanted to list the actual actions that this grants, i.e:

- `kinesis:DescribeStream`
- `kinesis:GetRecords`
- ...

Comment (iliapolo) [reply packages/@aws-cdk/aws-kinesis/README.md]: I would probably just create a list of common actions (those used in both read and write), and specifically mention those who are used only for either.

File: packages/@aws-cdk/aws-kinesis/lib/stream.ts [hunk 3]
```diff
@@ -163,7 +164,10 @@ abstract class StreamBase extends Resource implements IStream {
     return ret;
   }
 
-  private grant(grantee: iam.IGrantable, ...actions: string[]) {
+  /**
+   * Grant the indicated permissions on this key to the given IAM principal (Role/Group/User).
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-kinesis/lib/stream.ts]: Why do you say `permissions on this key`, rather than `permissions` on this stream?

**Comment by iliapolo** [comment]
@reillykw Notice the build is failing, i think you might need to update the integ expectations again.

**Comment by shivlaks** [comment]
> @reillykw Notice the build is failing, i think you might need to update the integ expectations again.

@reillykw you also may want to pull / merge from master before doing that as there are a bunch of other changes in the Kinesis module that might present conflicts (including the `aws-lambda-event-sources` integ test that's currently failing.

File: packages/@aws-cdk/aws-kinesis/README.md (hunk 1 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-kinesis/README.md]: updated

File: packages/@aws-cdk/aws-kinesis/lib/stream.ts (hunk 3 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-kinesis/lib/stream.ts]: updated

File: packages/@aws-cdk/aws-kinesis/lib/stream.ts [hunk 4]
```diff
@@ -132,7 +139,11 @@ abstract class StreamBase extends Resource implements IStream {
    * contents of the stream will also be granted.
    */
   public grantWrite(grantee: iam.IGrantable) {
-    const ret = this.grant(grantee, 'kinesis:DescribeStream', 'kinesis:PutRecord', 'kinesis:PutRecords');
+    const ret = this.grant(
+      grantee,
+      'kinesis:ListShards',
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-kinesis/lib/stream.ts:144]: Is `ListShards` needed for write? I know `DescribeStream` was here but do we know why?
Also - are we sure getting rid of it doesn't break anything? doesn't that mean that users who previously could describe a stream with just the write permissions won't be able to now?

Im ok with that, we just need to publish this as a breaking change in that case.

File: packages/@aws-cdk/aws-kinesis/lib/stream.ts [hunk 5]
```diff
@@ -152,10 +163,13 @@ abstract class StreamBase extends Resource implements IStream {
     const ret = this.grant(
       grantee,
       'kinesis:DescribeStream',
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-kinesis/lib/stream.ts:165]: I'd love to extract this array to `READ_OPERATIONS` and `WRITE_OPERATIONS` and use those constants in `grantRead` and `grantWrite` accordingly. 

Actually ideally the implementation of `grantReadWrite` would just be:

```typescript
this.grantRead()
this.grantWrite()
```

Is that possible? Not sure what the return value will be in such a case.

File: packages/@aws-cdk/aws-kinesis/README.md [hunk 6]
```diff
@@ -119,3 +122,65 @@ const importedStream = Stream.fromStreamAttributes(
   }
 );
 ```
+
+### Permission Grants
+
+IAM roles, users or groups which need to be able to work with Amazon Kinesis streams at runtime will should be granted IAM permissions. Any object that implements the `IGrantable`
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-kinesis/README.md:128]:
```suggestion
IAM roles, users or groups which need to be able to work with Amazon Kinesis streams at runtime should be granted IAM permissions. Any object that implements the `IGrantable`
```

File: packages/@aws-cdk/aws-kinesis/lib/stream.ts (hunk 4 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-kinesis/lib/stream.ts:144]: yes it is, I was advised to add it by @yasemin-amzn to align with the Kinesis Producer Library.

I included the rationale already for breaking change in the commit message... is there anything to do here?

File: packages/@aws-cdk/aws-kinesis/README.md (hunk 6 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-kinesis/README.md]: updated

File: packages/@aws-cdk/aws-kinesis/lib/stream.ts (hunk 5 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-kinesis/lib/stream.ts]: took a shot at it, let me know what you think

File: packages/@aws-cdk/aws-kinesis/lib/stream.ts (hunk 4 above)
Comment (iliapolo) [reply packages/@aws-cdk/aws-kinesis/lib/stream.ts]: Whoops - didn't notice the breaking changes notice. All good 👍

File: packages/@aws-cdk/aws-kinesis/lib/stream.ts [hunk 7]
```diff
@@ -149,13 +164,7 @@ abstract class StreamBase extends Resource implements IStream {
...
    */
   public grantReadWrite(grantee: iam.IGrantable) {
-    const ret = this.grant(
-      grantee,
-      'kinesis:DescribeStream',
-      'kinesis:GetRecords',
-      'kinesis:GetShardIterator',
-      'kinesis:PutRecord',
-      'kinesis:PutRecords');
+    const ret = this.grant(grantee, ...Array.from(new Set([...READ_OPERATIONS, ...WRITE_OPERATIONS])));
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-kinesis/lib/stream.ts:167]: This is what we usually do i think: maybe something like this? https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-s3/lib/bucket.ts#L524 ?

Comment (iliapolo) [reply packages/@aws-cdk/aws-kinesis/lib/stream.ts:167]: Oh its because of the overlapping action. its ok i guess.

**Pull request merged**